src/main/kotlin/org/jmailen/gradle/kotlinter/pluginapplier/AndroidSourceSetApplier.kt

```diff
@@ -14,18 +14,21 @@ import org.jmailen.gradle.kotlinter.id
 internal object AndroidSourceSetApplier : SourceSetApplier {
 
     override fun applyToAll(project: Project, action: SourceSetAction) {
-        val android = project.extensions.findByName("android")
-        (android as? BaseExtension)?.let {
-            it.sourceSets.all { sourceSet ->
-                val id = sourceSet.name.id
-                action(id, project.provider { getKotlinFiles(project, sourceSet) })
-            }
+        val android = project.extensions.findByName("android") as? BaseExtension ?: return
+        android.sourceSets.configureEach { sourceSet ->
+            val id = sourceSet.name.id
+            action(id, project.provider { getKotlinFiles(project, sourceSet) })
         }
     }
 
-    private fun getKotlinFiles(project: Project, sourceSet: AndroidSourceSet) = sourceSet.java.srcDirs.map { dir ->
-        project.fileTree(dir) { it.include("**/*.kt") }
-    }.reduce { merged: FileTree, tree: ConfigurableFileTree ->
-        merged + tree
+    private fun getKotlinFiles(project: Project, sourceSet: AndroidSourceSet): FileTree? {
+        val javaSources = sourceSet.java.srcDirs
+        val kotlinSources = runCatching { (sourceSet.kotlin as? com.android.build.gradle.api.AndroidSourceDirectorySet)?.srcDirs }
```

Comment:
Let's put a comment above this line noting why. Something like "detect kotlin source paths supported by AGP 7 and later"